### DISCUSSION TITLE
Import Custom Vue Component with Dependency

### DISCUSSION THREAD

**Discussion by ShadowBolt09** [zauberzeug/nicegui#2928] [Q&A]
### Question

Hello there, :smile:

I need some guidance on how to import a `Vue` component into my `nicegui` project, particularly speaking, `Vue-Ganttastic`. I have been following [this guide](https://github.com/zauberzeug/nicegui/tree/main/examples/custom_vue_component/) to import custom `Vue` components to no success.

---

Here is the most minimal reproducible code example I could think of:

:page_facing_up: `ganttastic.py`
```python
from typing import Callable, Optional
from nicegui.element import Element
class Ganttastic(Element, component='ganttchart.js'):
    def __init__(self, title: str, *, on_change: Optional[Callable] = None) -> None:
        super().__init__()
        # self.set_props(chart_start=chart_start, chart_end=chart_end, rows=rows)
        self.on_change = on_change
```

:page_facing_up: `main.py` _(ui.add_body_html was used in an attempt to use CDN links with no success)_
```python
from nicegui import ui
ui.add_body_html('''
<script src="https://unpkg.com/vue@3.2.31/dist/vue.global.prod.js"></script>
<script src="https://unpkg.com/dayjs@1.8.21/dayjs.min.js"></script>
''')
from ganttastic import Ganttastic
# Define the rows for the Gantt chart with actual data
rows_data = [
    # Define your rows data here according to Vue-Ganttastic's expected format
]
Ganttastic(title='Ganttastic')
ui.run()
```

:page_facing_up: `ganttchart.js` _(this code is insanely huge so I will share with you the first 9 lines of it, the only parts that matter are the imports)_
```javascript
import A from "dayjs";
import { inject as ue, computed as R, defineComponent as Q, openBlock as O, createElementBlock as M, Fragment as ee, renderList as te, unref as C, normalizeStyle as L, renderSlot as S, createElementVNode as G, toDisplayString as q, createTextVNode as ne, createCommentVNode as V, toRefs as re, ref as Y, watch as ce, nextTick as Le, createBlock as J, Teleport as et, createVNode as Be, Transition as tt, withCtx as N, getCurrentScope as nt, onScopeDispose as rt, getCurrentInstance as at, onMounted as Se, useSlots as ot, provide as le, normalizeClass as be, withModifiers as it, mergeProps as st, TransitionGroup as lt } from "vue";
var K = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, Ye = { exports: {} };
(function(e, s) {
  (function(r, n) {
    e.exports = n();
  })(K, function() {
    var r = "day";
    return function(n, p, l) {
```

This is what happens when I try to run the program:
![image](https://github.com/zauberzeug/nicegui/assets/60474869/78faccd4-9a75-4715-b70f-e89b95831f02)

And when I remove the `dayjs` `import` I get the same error but with `Vue`.

---

Run with:
```powershell
python main.py
```

Pardon me if this is a novice question to ask. Hope someone can help me out.
Thank you so much in advance for reading. Any guidance would be appreciated. :pray:

**Comment by falkoschindler** [comment]
Hi @ShadowBolt09,

Please have a look at discussion https://github.com/zauberzeug/nicegui/discussions/1454. We're still working on an easy way to integrate such third-party JavaScript and Vue libraries. Hopefully we'll have a solution for the upcoming major release of version [2.0](https://github.com/zauberzeug/nicegui/milestone/82).

**Comment by traverseda** [comment]
The linked discussion from last year isn't available. Did this feature get completed? I'm very interested in integrating [vue-term](https://github.com/gwangyi/vue-term) into my ROS application.

**Reply by evnchn** [reply]
Hmm no AFAIK. Adding third party JS and Vue libraries remains a DIY process, and there was never a turnkey solution. 

I mean, look at the to-do features in https://github.com/orgs/zauberzeug/projects/6, and you will find that even some Quasar elements are not covered by NiceGUI. 

So, if you want to add vue-term, you will most likely have to do it yourself. And, don't forget to PR back to NiceGUI if you got it working 💪

**Reply by falkoschindler** [reply]
Oh wow, somehow discussion #1454 got deleted. 😕

Anyway. If vue-term = xtermjs + vue.js, it's probably easier to integrate xtermjs into NiceGUI. A starting point might be [this section](https://nicegui.io/documentation/section_configuration_deployment#custom_vue_components) and the examples [custom Vue component](https://github.com/zauberzeug/nicegui/tree/main/examples/custom_vue_component), [signature pad](https://github.com/zauberzeug/nicegui/tree/main/examples/signature_pad), [node module](https://github.com/zauberzeug/nicegui/tree/main/examples/node_module_integration).

**Comment by evnchn** [comment]
@ShadowBolt09 I don't know if you are still here, but I just have some new insights. 

The error you got, it is caused by missing declaration in the importmap, and then `import A from "dayjs";` throws up. 

Faced a similar issue implementing https://github.com/zauberzeug/nicegui/pull/4507, where ui.aggrid would break because I put the script at the top before the importmap. Specifically https://github.com/zauberzeug/nicegui/pull/4507/commits/0bf680709d1f5c62bbf2aac7259605499bfa543f

So, the name of the game is likely to get dayjs to be inside the import map, so that you can safely import it.
